dialog/file.go

```diff
@@ -326,48 +328,49 @@ func (f *fileDialog) loadFavorites() {
 }
 
 func (f *fileDialog) refreshDir(dir fyne.ListableURI) {
-	f.files.Objects = nil
+	f.data = nil
 
 	files, err := dir.List()
 	if err != nil {
 		fyne.LogError("Unable to read ListableURI "+dir.String(), err)
 		return
 	}
 
-	var icons []fyne.CanvasObject
+	var icons []fyne.URI
 	parent, err := storage.Parent(dir)
 	if err != nil && err != repository.ErrURIRoot {
 		fyne.LogError("Unable to get parent of "+dir.String(), err)
 		return
 	}
 	if parent != nil && parent.String() != dir.String() {
-		fi := &fileDialogItem{picker: f, name: "(Parent)", location: parent, dir: true}
-		fi.ExtendBaseWidget(fi)
-		icons = append(icons, fi)
+		icons = append(icons, parent)
 	}
 
 	for _, file := range files {
 		if !f.showHidden && isHidden(file) {
 			continue
 		}
 
-		listable, err := storage.CanList(file)
+		listable, err := storage.ListerForURI(file)
 		if f.file.isDirectory() && err != nil {
 			continue
-		} else if err == nil && listable { // URI points to a directory
-			icons = append(icons, f.newFileItem(file, true)) // Pass the listable URI to avoid doing the same check in FileIcon
+		} else if err == nil { // URI points to a directory
+			icons = append(icons, listable)
 		} else if f.file.filter == nil || f.file.filter.Matches(file) {
-			icons = append(icons, f.newFileItem(file, false))
+			icons = append(icons, file)
 		}
 	}
+	f.data = icons
 
-	f.files.Objects = icons
 	f.files.Refresh()
 	f.filesScroll.Offset = fyne.NewPos(0, 0)
 	f.filesScroll.Refresh()
 }
 
 func (f *fileDialog) setLocation(dir fyne.URI) error {
+	if f.selectedID > -1 {
+		f.files.(interface{ Unselect(int) }).Unselect(f.selectedID)
```

Comment:
You can probably move this interface into its own type and add `.Refresh()` you use that instead of `fyne.Widget` in the fileDialog struct and avoid casting the value here.

Reply:
Great idea thanks. Though I extended `fyne.Widget` so it could be passed to container constructor too.